### PR TITLE
[FLINK-16996][table] Refactor planner and connectors to use new data structures

### PR DESCRIPTION
**Pull request by wuchong**
## What is the purpose of the change

Refactors existing code to use the new data structures interfaces.

## Brief change log

The commints in the order:
- [table-common] Add necessary methods to internal data structures
- [table-common] Add binary implementations of internal data structures
- [table-runtime-blink] Implement all the data structures and serializers around RowData
- [table-runtime-blink] Remove legacy data formats (BaseRow)
- [table-blink] Refactor planner and runtime to use new data structures
- [python] Refactor pyflink to use new data structures
- [parquet] Refactor parquet connector to use new data structures
- [orc] Refactor ORC connector to use new data structures
- [hive] Refactor Hive connector to use new data structures

Some notable changes: 
- In code generation, we hard cast `StringData` to `BinaryStringData`. This makes the code generator easily to generate opeartions based on string. The same to `RawValueData`. 
- Most methods of `Decimal` have been moved to `DecimalDateUtil`. So I also updated the code generation logic.
- Remove `RecordEqualiser#equalsWithoutHeader` interface. This method is used less and can be replaced by `RecordEqualiser#equals`. This can also avoid add `equalsWithoutHeader` method to the public API `GenericRowData`.

## Verifying this change

This change is covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: yes
  - The runtime per-record code paths (performance sensitive): yes
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

**Comment by wuchong** [comment]
Hi @dianfu , could you help to review the python part? `RowData` and `ArrayDat` don't implement `TypeGetterSetter`, so I have to refactor the `ArrowFieldWriter` implementations.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/data/binary/BinaryStringDataUtil.java [hunk 1]
```diff
@@ -134,49 +136,49 @@
...
-	 * @return Decimal value if the parsing was successful, or null if overflow
+	 * @return DecimalData value if the parsing was successful, or null if overflow
 	 * @throws NumberFormatException if the parsing failed.
 	 */
-	public static Decimal toDecimal(BinaryString str, int precision, int scale) {
+	public static DecimalData toDecimal(BinaryStringData str, int precision, int scale) {
 		str.ensureMaterialized();
 
-		if (precision > Decimal.MAX_LONG_DIGITS || str.getSizeInBytes() > Decimal.MAX_LONG_DIGITS) {
+		if (DecimalDataUtils.isByteArrayDecimal(precision) || DecimalDataUtils.isByteArrayDecimal(str.getSizeInBytes())) {
```
Comment (KurtYoung) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/data/binary/BinaryStringDataUtil.java:171]: Is this right? why we check both `precision` and `sizeInBytes` with the same method?

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/data/util/MapDataUtil.java [hunk 2]
```diff
@@ -16,44 +16,34 @@
...
-
-	public Map<Object, Object> getMap() {
-		return map;
+public final class MapDataUtil {
+
+	/**
+	 * Converts a {@link MapData} into Java {@link Map}, the keys and values of the Java map
+	 * still holds objects of internal data structures.
+	 */
+	public static Map<Object, Object> convertToJavaMap(
```
Comment (KurtYoung) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/data/util/MapDataUtil.java:37]: only used by tests

Comment (wuchong) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/data/util/MapDataUtil.java:37]: This is also used in code generation: 
https://github.com/wuchong/flink/blob/744983929e464758baad98bf73cba9601eac48c5/flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/ScalarOperatorGens.scala#L2179

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/data/binary/BinaryStringDataUtil.java (hunk 1 above)
Comment (wuchong) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/data/binary/BinaryStringDataUtil.java:171]: cc @JingsongLi , do you know why we check both of them?

**Comment by dianfu** [comment]
@wuchong The python part LGTM.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/data/GenericRowData.java [hunk 3]
```diff
@@ -186,6 +187,20 @@ public RowData getRow(int pos, int numFields) {
...
 
+	@Override
+	public String toString() {
+		StringBuilder sb = new StringBuilder();
+		sb.append(kind.shortString()).append("(");
+		for (int i = 0; i < fields.length; i++) {
+			if (i != 0) {
+				sb.append(",");
+			}
+			sb.append(StringUtils.arrayAwareToString(fields[i]));
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/data/GenericRowData.java:215]: Use `org.apache.flink.table.utils.EncodingUtils#objectToString`

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/AsyncTableFunction.java [hunk 4]
```diff
@@ -55,7 +55,7 @@
  * runtime.
  *
  * <p>By default the result type of an evaluation method is determined by Flink's type extraction
- * facilities. Currently, only support {@link org.apache.flink.types.Row} and {@code BaseRow} as
+ * facilities. Currently, only support {@link org.apache.flink.types.Row} and {@code RowData} as
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/AsyncTableFunction.java]: use: `{@link RowData}`

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/utils/LogicalTypeUtils.java [hunk 5]
```diff
@@ -36,6 +43,54 @@ public static LogicalType removeTimeAttributes(LogicalType logicalType) {
...
+				return MapData.class;
+			case ROW:
+				return RowData.class;
+			case BINARY:
+			case VARBINARY:
+				return byte[].class;
+			case RAW:
+				return RawValueData.class;
+			default:
+				throw new UnsupportedOperationException("Not support type: " + type);
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/utils/LogicalTypeUtils.java:92]: nit: `Unsupported type:`

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/utils/LogicalTypeUtils.java [hunk 6]
```diff
@@ -36,6 +43,54 @@ public static LogicalType removeTimeAttributes(LogicalType logicalType) {
 		return logicalType.accept(TIME_ATTRIBUTE_REMOVER);
 	}
 
+	/**
+	 * Get internal(sql engine execution data formats) conversion class for {@link LogicalType}.
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/utils/LogicalTypeUtils.java]: nit: `Returns the conversion class for the given {@link LogicalType} that is used by the table runtime.`

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/utils/LogicalTypeUtils.java [hunk 7]
```diff
@@ -36,6 +43,54 @@ public static LogicalType removeTimeAttributes(LogicalType logicalType) {
...
+	 */
+	public static Class<?> internalConversionClass(LogicalType type) {
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/logical/utils/LogicalTypeUtils.java:51]: nit: `toInternalConversionClass`

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/data/binary/BinaryFormat.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+ */
+
+package org.apache.flink.table.data.binary;
+
+import org.apache.flink.core.memory.MemorySegment;
+
+/**
+ * Binary format spanning {@link MemorySegment}s.
+ */
+public interface BinaryFormat {
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/data/binary/BinaryFormat.java:28]: Missing `@Internal`

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/data/binary/LazyBinaryFormat.java [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+ * <ul>
+ *     <li>Binary form</li>
+ *     <li>Java object form</li>
+ *     <li>Binary and Java object both exist</li>
+ * </ul>
+ *
+ * <p>It can lazy the conversions as much as possible. It will be converted into required form
+ * only when it is needed.
+ */
+public abstract class LazyBinaryFormat<T> implements BinaryFormat {
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/data/binary/LazyBinaryFormat.java:57]: Missing `@Internal`

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/BinarySegmentUtils.java [hunk 10]
```diff
@@ -0,0 +1,1201 @@
...
+
+import static org.apache.flink.core.memory.MemoryUtils.UNSAFE;
+import static org.apache.flink.table.data.binary.BinaryFormat.HIGHEST_FIRST_BIT;
+import static org.apache.flink.table.data.binary.BinaryFormat.HIGHEST_SECOND_TO_EIGHTH_BIT;
+
+/**
+ * Utilities for binary data segments which heavily uses {@link MemorySegment}.
+ */
+@Internal
+public class BinarySegmentUtils {
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/BinarySegmentUtils.java:49]: Put this under `org.apache.flink.table.data.binary`. Make the class `final` with a private default constructor.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/MurmurHashUtils.java [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.core.memory.MemorySegment;
+
+import static org.apache.flink.core.memory.MemoryUtils.UNSAFE;
+
+/**
+ * Murmur Hash. This is inspired by Guava's Murmur3_32HashFunction.
+ */
+@Internal
+public final class MurmurHashUtils {
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/MurmurHashUtils.java:30]: put under `org.apache.flink.table.data.binary` add private default constructor

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/StringUtf8Utils.java [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+import java.util.Arrays;
+
+import static org.apache.flink.table.utils.BinarySegmentUtils.allocateReuseBytes;
+import static org.apache.flink.table.utils.BinarySegmentUtils.allocateReuseChars;
+
+/**
+ * Utilities for String UTF-8.
+ */
+@Internal
+public class StringUtf8Utils {
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/StringUtf8Utils.java:34]: put under `org.apache.flink.table.data.binary` make final with private default constructor

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/data/TypedSetters.java [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.data.binary.BinaryRowData;
+
+/**
+ * Provide type specialized setters to reduce if/else and eliminate box and unbox. This is mainly
+ * used on the binary format such as {@link BinaryRowData}.
+ */
+@Internal
+public interface TypedSetters {
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/data/TypedSetters.java:29]: If this is mainly used for binary formats, put it in the binary package.

**Comment by twalthr** [comment]
Thanks @wuchong for this massive PR. I took a look at the classes in `table-common`. Apart from my comments, they look good to me.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/CodeGenUtils.scala [hunk 14]
```diff
@@ -132,6 +136,15 @@ object CodeGenUtils {
    * Returns a term for representing the given class in Java code.
    */
   def typeTerm(clazz: Class[_]): String = {
+    if (clazz == classOf[StringData]) {
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/CodeGenUtils.scala:139]: Could this put CodeGen at risk of missing cast?
If possible, I prefer cast in accessing its methods.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/CodeGeneratorContext.scala [hunk 15]
```diff
@@ -678,7 +678,7 @@ class CodeGeneratorContext(val tableConfig: TableConfig) {
 
       case None =>
         val term = newName("typeSerializer")
-        val ser = InternalSerializers.create(t, new ExecutionConfig)
+        val ser = InternalSerializers.createInternalSerializer(t, new ExecutionConfig)
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/CodeGeneratorContext.scala:681]: `createInternalSerializer` a little redundant

**Comment by wuchong** [comment]
Thanks @KurtYoung @dianfu @twalthr for the quickly reviewing. I have addressed all the comments. 

Hi Timo, currently, it's hard to make all classes under `binary.` to be package-visible, because they will be used in serializers and code generation. But I think `MurmurHashUtils` and `StringUtf8Utils` can be, so I updated them.

**Comment by JingsongLi** [comment]
There are still 100+ BaseRow in codes (variable name or something else), you can modify them all.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/CodeGenUtils.scala (hunk 14 above)
Comment (JingsongLi) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/CodeGenUtils.scala:139]: Consider `RowData` and others, using `StringData` more reasonable to me.

Comment (wuchong) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/CodeGenUtils.scala:139]: There are too many calls on the `BinaryStringData` in code generation now, if we use StringData in code generation, we have to refactor a lot of codes. And we don't benefit much from this effort. I created FLINK-17437 to track this, we can refactor this in the future.

**Comment by wuchong** [comment]
Thanks for the reviewing @JingsongLi . I have addressed the comments and renamed lagecy field names and method using BaseRow, SqlTimestamp and BinaryGeneric.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/data/binary/BinaryStringDataUtil.java (hunk 1 above)
Comment (JingsongLi) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/data/binary/BinaryStringDataUtil.java:171]: The case is, precision is 10 (less than `MAX_LONG_DIGITS`), but the string data is more than 10 precision maybe, we can use big decimal to convert, and then remove the extra decimals according to the precision.

**Comment by wuchong** [comment]
Thanks all for the reviewing. I have rebased the commits. Will merge this once builds passed.

**Comment by JingsongLi** [comment]
Hi @wuchong , do you want to squash or not?  Here is my question:
Does it need to be able to compile for each commit to the master?

**Comment by wuchong** [comment]
Hi @JingsongLi , I want to keep the splitted commits. From my point of view, squashing such a large PR into one commit is not good.

**Comment by JingsongLi** [comment]
> Hi @JingsongLi , I want to keep the splitted commits. From my point of view, squashing such a large PR into one commit is not good.

So the answer is we can push commit with broken compilation.

**Comment by wuchong** [comment]
@JingsongLi , as far as I know, the community doesn't have a rule to make sure independent commits can pass build, but there is a rule to separate commits if it is a big commit:

https://flink.apache.org/contributing/code-style-and-quality-pull-requests.html#separate-refactoring-cleanup-and-independent-changes

**Comment by wuchong** [comment]
Travis is passed: https://travis-ci.org/github/wuchong/flink/builds/680843224

**Comment by JingsongLi** [comment]
@wuchong go you.

**Pull request closed without merging**